sdks/go/pkg/beam/io/bigtableio/bigtable.go

```diff
@@ -0,0 +1,278 @@
+func (m *Mutation) WithGroupKey(key string) *Mutation {
+	m.GroupKey = key
+	return m
+}
+
+// Write writes the elements of the given PCollection<bigtableio.Mutation> to bigtable.
+func Write(s beam.Scope, project, instanceID, table string, col beam.PCollection) {
+	t := col.Type().Type()
+	err := mustBeBigtableioMutation(t)
+	if err != nil {
+		panic(err)
+	}
+
+	s = s.Scope("bigtable.Write")
+
+	pre := beam.ParDo(s, addGroupKeyFn, col)
+	post := beam.GroupByKey(s, pre)
+	beam.ParDo0(s, &writeFn{Project: project, InstanceID: instanceID, TableName: table, Type: beam.EncodedType{T: t}}, post)
+}
+
+// WriteBatch writes the elements of the given PCollection<bigtableio.Mutation> 
+// to bigtable using bigtable.ApplyBulk().
+// For the underlying bigtable.ApplyBulk function to work properly 
+// the maximum number of operations per bigtableio.Mutation of the input
+// PCollection must be given (maxOpsPerMutation). This is necessary due to 
+// the maximum amount of mutations allowed per bulk operation (100,000),
+// see https://cloud.google.com/bigtable/docs/writes#batch for more.
+func WriteBatch(s beam.Scope, project, instanceID, table string, maxOpsPerMutation uint, col beam.PCollection) {
+	t := col.Type().Type()
+	err := mustBeBigtableioMutation(t)
+	if err != nil {
+		panic(err)
+	}
+
+	if maxOpsPerMutation == 0 {
+		panic("maxOpsPerMutation must not be 0")
+	}
+
+	if maxOpsPerMutation > 100000 {
+		panic("maxOpsPerMutation must not be greater than 100,000, see https://cloud.google.com/bigtable/docs/writes#batch")
+	}
+
+	s = s.Scope("bigtable.WriteBatch")
+
+	pre := beam.ParDo(s, addGroupKeyFn, col)
+	post := beam.GroupByKey(s, pre)
+	beam.ParDo0(s, &writeBatchFn{Project: project, InstanceID: instanceID, TableName: table, MaxOpsPerMutation: maxOpsPerMutation, Type: beam.EncodedType{T: t}}, post)
+}
+
+func addGroupKeyFn(mutation Mutation) (int, Mutation) {
+	if mutation.GroupKey != "" {
+		return hashStringToInt(mutation.GroupKey), mutation
+	}
+	return 1, mutation
+}
+
+func hashStringToInt(s string) int {
+	h := fnv.New32a()
+	h.Write([]byte(s))
+	return int(h.Sum32())
+}
+
+func mustBeBigtableioMutation(t reflect.Type) error {
+	if t != reflect.TypeOf(Mutation{}) {
+		return fmt.Errorf("type must be bigtableio.Mutation but is: %v", t)
+	}
+	return nil
+}
+
+type writeFn struct {
+	// Project is the project
+	Project string `json:"project"`
+	// InstanceID is the bigtable instanceID
+	InstanceID string `json:"instanceId"`
+	// Client is the bigtable.Client
+	Client *bigtable.Client `json:"client"`
+	// TableName is the qualified table identifier.
+	TableName string `json:"tableName"`
+	// Table is a bigtable.Table instance with an eventual open connection
+	Table *bigtable.Table `json:"table"`
+	// Type is the encoded schema type.
+	Type beam.EncodedType `json:"type"`
+}
+
+func (f *writeFn) StartBundle() error {
+	var err error
+	f.Client, err = bigtable.NewClient(context.Background(), f.Project, f.InstanceID)
```

Comment:
While these appear to be intended for batch usage, runners can re-use "plan" instances, and bundles can be very very small if decided by the runner. Consider moving client and table set up to a Setup method to avoid constantly connecting and reconnecting.

If bigtable Clients and Tables are threadsafe, it would also be worth considering to move their initialization to a package level cache for re-use among instances.

Reply:
Moved Client initialisation to Setup and Client closing to Teardown function.
Thanks for the optimisation!